File: src/python/finished/number_of_laser_beams_in_a_bank.py
from typing import List


# https://leetcode.com/problems/number-of-laser-beams-in-a-bank/


class Solution:
    def numberOfBeams(self, bank: List[str]) -> int:
        prev_device_cnt = 0
        beams = 0
        for row in bank:
            device_cnt = row.count("1")
            if device_cnt == 0:
                continue

            beams += device_cnt * prev_device_cnt
            prev_device_cnt = device_cnt

        return beams

File: src/rust/finished/number_of_laser_beams_in_a_bank.rs
impl Solution {
    pub fn number_of_beams(bank: Vec<String>) -> i32 {
        bank.iter()
            .fold((0, 0), |(beams, pcnt), row| {
                let laser_cnt = row
                    .chars()
                    .fold(0, |prev, ch| prev + if ch == '1' { 1 } else { 0 });
                if laser_cnt == 0 {
                    return (beams, pcnt);
                }

                (beams + pcnt * laser_cnt, laser_cnt)
            })
            .0
    }
}
